### ISSUE TITLE
Using CreateProcessA in VBA not opening an MPV console

### ISSUE THREAD

**Issue by 99bobster99** [mpv-player/mpv#8664]
Provide following Information:
- mpv version (mpv-x86_64-20210228-git-d1be8bb.7z    &     mpv-x86_64-20210316-git-5824d9f.7z)
- Windows Version (Windows 7 x64 Ultimate SP1)
- Source of the mpv binary (https://sourceforge.net/projects/mpv-player-windows/files/64bit/)
- If known which version of mpv introduced the problem (the last 2 versions I've checked)
- Possible screenshot or video of visual glitches (none)

If you're not using git master or the latest release, update.
Releases are listed here: https://github.com/mpv-player/mpv/releases   (done, tried latest as well)

### Reproduction 

I tried all combinations of command line options for mpv, that involved the pipe inputs and tested them in a CMD prompt first to make sure they worked. 

The one I am using was the most simplest one I could use that worked fine when using the CMD prompt. There was obviously no video being displayed, but the MPV console did appear. I also used the "--ontop" option to try to force the console to the top should it have been created.

### Expected behavior

Run the code below, in VBA, using windows pipe between FFMMPEG and STDIN of mpv. MPV player should open as a normal console.

For both tests below, the FFMPEG command used was;

        FFMPEG = "c:\projects\ffmpeg -loglevel quiet -nostats -i c:\projects\testvideo.mkv -c:v copy -c:a copy -f mpegts pipe:1"

`
        strProgram = "c:\test1\mpv - --no-config --idle"

        FillMemory piProcInfo, Len(piProcInfo), 0&
        FillMemory siStartInfo, Len(siStartInfo), 0&

        siStartInfo.dwFlags = STARTF_USESTDHANDLES
        siStartInfo.hStdInput = hStdIn      ( <- other end of this pipe is uninheritable)
        siStartInfo.hStdOutput = hStdOut     ( <- other end of this pipe is uninheritable)
        siStartInfo.hStdError = hStdErr     ( <- other end of this pipe is uninheritable)
        siStartInfo.wShowWindow = SW_SHOWNORMAL
         
        ok = CreateProcessA(vbNullString, strProgram, ByVal 0&, ByVal 0&, 1&, 0, ByVal 0&, vbNullString, siStartInfo, piProcInfo)  

`

During testing, I tried the same code as above, but using VLC and it worked as expected (console opened for VLC and piped output from FFMPEG played in VLC). 

This is the code used for the VLC testing;

`
        strProgram = "c:\test1\vlc -"

        FillMemory piProcInfo, Len(piProcInfo), 0&
        FillMemory siStartInfo, Len(siStartInfo), 0&

        siStartInfo.dwFlags = STARTF_USESTDHANDLES
        siStartInfo.hStdInput = hStdIn      ( <- other end of this pipe is uninheritable)
        siStartInfo.hStdOutput = hStdOut     ( <- other end of this pipe is uninheritable)
        siStartInfo.hStdError = hStdErr     ( <- other end of this pipe is uninheritable)
        siStartInfo.wShowWindow = SW_SHOWNORMAL
         
        ok = CreateProcessA(vbNullString, strProgram, ByVal 0&, ByVal 0&, 1&, 0, ByVal 0&, vbNullString, siStartInfo, piProcInfo)  

`

### Actual behavior

During the testing with mpv player, the "mpv.exe" process is displayed in the Windows Task Manager (size  = 13,356k), but there is no console created. As I write frames to mpv, using "WriteFile", I see the size of the mpv process increase in Task Manager but no console is being displayed.

The Read and Write file code is shown below;

```
Do While (-1)

    If Len(ConsoleOutBuffer) < 10000000 Then
Retry1:
        ok = PeekNamedPipe(h1OutReadPipe, ByVal 0&, 0&, ByVal 0&, cbSizeToRead, ByVal 0&)
        If Not ok Then GoTo Error1
        
        If ok And cbSizeToRead > 0 Then
            ConsoleOutBufferTmp = Space(cbSizeToRead)
            ok = ReadFile(h1OutReadPipe, ByVal ConsoleOutBufferTmp, cbSizeToRead, cbSizeRead, 0&)
            If Not ok Then GoTo Error1
            ConsoleOutBuffer = ConsoleOutBuffer & ConsoleOutBufferTmp
            bCheckForEOF = True
            iEndCount = 0
        ElseIf bCheckForEOF Then
            iEndCount = iEndCount + 1
            If iEndCount > 1000 Then
                Exit Do
            End If
            GoTo Retry1
        End If
    Else
    
        ok = WriteFile(h2InWritePipe, ByVal ConsoleOutBuffer, Len(ConsoleOutBuffer), cbSizeWritten, 0&)
        If Not ok Then Exit Do
        
        ConsoleOutBuffer = ""
    End If
        
    DoEvents
    
Loop
```

### Log file

Log file created, link is below;
[https://0x0.st/-N_5.txt](url)

**Comment by sfan5** [comment]
I don't know how VBA picks the executable to run but try explicitly invoking `mpv.com` since that deals with console stuff:

https://github.com/mpv-player/mpv/wiki/FAQ#On_Windows_why_does_mpvexe_not_attach_to_the_console_and_what_does_mpvcom_do

**Comment by 99bobster99** [comment]
Calling `mpv.com` instead of `mpv.exe` , does open a console now, but there is no secondary video console opening to play the stdin video stream (both `mpv.com` and `mpv.exe` are running in Task Manager). Entering `mpv.com - --no-config --idle --no-input-terminal` in a newly opened command window, shows  `[file] Reading from stdin... Failed to recognize file format.` I'm wondering if mpv needs to see the stream present before starting, even though mpv is still running and looks like it is waiting for a stream to start coming in on stdin?

**Comment by avih** [comment]
Launching a process and figuring how to redirect stdin is outside the scope of mpv.

You can look at `mpv.com` source code, which does the same - it redirect the stdin of `mpv.com` to `mpv.exe`:
https://github.com/mpv-player/mpv/blob/master/osdep/win32-console-wrapper.c

**Comment by 99bobster99** [comment]
I think the problem is that `mpv.exe` is not allowing access, since the behavior is exactly the same when using `mpv.com` or my application. I really wanted to use mpv over vlc for my project. Please let me know if any other ideas come to mind.

**Comment by avih** [comment]
I don't understand your exact problem. Are you saying that mpv on windows cannot play a file from stdin?

Because this works for me (i.e. the video plays):
`mpv.com - < c:\whatever\myvideo.mkv`

and if you have a `cat` program (like in busybox-w32), then this also works for me from its shell:
`cat c:/whatever/myvideo.mkv | mpv.com -` 

And as I mentioned, mpv.com does redirect stdin to mpv.exe, and I posted a link to the source code.

If that's not what you asked, then please try to describe the issue exactly, in plain `cmd.exe` shell, without additional programming languages.

**Comment by avih** [comment]
> in a newly opened command window, shows `[file] Reading from stdin... Failed to recognize file format`

Also, consider the possibility that mpv really does not recognize the video format which you configured ffmpeg to output.

To test it, write the output from ffmpeg to a normal file, e.g. `ffmpeg.out`, and then try to play it with `mpv - < ffmpeg.out`.

**Comment by 99bobster99** [comment]
When using plain `cmd.exe` shell, the ffmpeg outputted video from `ffmpeg -loglevel quiet -nostats -i testvideo.mkv -c:v copy -c:a copy -f mpegts ffmpeg.out` played fine when running `mpv - < ffmpeg.out`. 

When `mpv.exe` is spawned from a `createprocess` command (same command `mpv - < ffmpegout` , without redirecting its stdin programmatically, the video plays normally inside a newly created console - as expected). 

The problem I'm seeing is that when `mpv.exe` is spawned from a `createprocess` command (using `mpv - --idle` , with its stdin redirected, the video does not play at all, since no video console is created. `mpv.exe` is shown in task manager and is accepting the stdin video stream, it's just not showing the video. This same stdin redirection test, using `vlc -` works perfectly fine. This tells me there is nothing wrong with the redirection code, and `mpv.exe` is capable of creating a console. `mpv.exe` is just not displaying the video.

**Comment by avih** [comment]
I'm not sure I understand your terminology:

> the video plays normally inside a newly created console
Is that windows console (like for cmd.exe)? video output window? something else?

> ... since no video console is created
Is that windows console? video output window? something else?

>  This tells me there is nothing wrong with the redirection code
Maybe, but personally I wouldn't make that assumption. It's entirely possible that there are other factors at play here, and your code happens to work well for some combinations of conditions (which happen with vlc), but not generically enough so that mpv falls between the cracks in your code. I haven't looked at your code so I just don't know.

> and mpv.exe is capable of creating a console
what "console" is that, and how is it relevant at all to your issue? as far as I know, all programs have stdin, regardless if they have or don't have a windows console.

> mpv - --idle
I don't think this instructs mpv to wait for video in stdin. As far as I know it means that mpv doesn't exit once playback is complete (or doesn't close right away if no file is provided). Not 100% sure though. You should check the docs.

Bottom line, other programs like cmd.exe, or busybox-w32 shell, or git-for-windows shell, and probably others can pipe from some program to stdin of mpv without knowing anything in advance about mpv. This tells ME that it is possible to do that in a generic way, but your code doesn't manage to do that for some reason.

**Comment by 99bobster99** [comment]
Sorry, for the terminology used! 

`the video plays normally inside a newly created console`  and  `... since no video console is created`
are the video output window

`This tells me there is nothing wrong with the redirection code`
There aren't many options for the `CreateProcess` command, I tried quite a lot of them and I never get a video output window. Do you have an example of what the `CreateProcess` code should look like to properly spawn `mpv.exe`? It could be there is a combination I have't tried yet?? 

`mpv - --idle`
I added the "--idle" because without it `mpv.exe` doesn't launch from a `CreateProcess` call or from plain `cmd.exe`. 

It must come down to getting the right attributes entered for the `CreateProcess` operation, I'm guessing. 

Keep in mind, the CreateProcess properly produces a video output window when using the `mpv - < ffmpegout` command. This is a good sign, it must be the redirection? But it is quite simple redirection that I'm using, I use it all the time, not sure why there is a problem (i.e. why I assumed since vlc works mpv should as well)

**Comment by avih** [comment]
For reference, this works from `cmd.exe` and from busybox-w32 shell, with both `mpv.com` and `mpv.exe` (video window opens, and the video plays in mpv):
`ffmpeg -f lavfi -i testsrc=duration=10:size=1280x720:rate=30 -c:v h264 -f mpegts - | mpv.com --no-config -`

So other programs can clearly connect stdout of ffmpeg to stdin of mpv correctly.
Your program can't. It's most probably a bug in your program.

**Comment by avih** [comment]
> I added the "--idle" because without it mpv.exe doesn't launch from a CreateProcess call OR from plain cmd.exe

I'm quite positive it does launch. It just closes immediately if you asked it to read from stdin but there's nothing there.

**Comment by 99bobster99** [comment]
Yeah I agree, but what about spawning mpv.exe from a `CreateProcess` command? I have no issues when running `mpv.exe` when using `cmd.exe`, it always works 100%, flawlessly, since that's the way I have been using it in the past. I'm trying to integrate it into my homebrew IPC program. Can you test if `mpv.exe` works from a `CreateProcess`, on your end, using redirection? I can send you my code, it's just an xls file for now until I prove out the theory.

**Comment by avih** [comment]
I gave you the link earlier - the source code of `mpv.com` - which launches `mpv.exe` using `CreateProcessW`, and controls its stdin.

**Comment by avih** [comment]
I don't think the A/W is relevant. It's probably only to be able to give unicode file names as arguments. But you should check the MSDN docs if there are subtle diffetences between the A/W versions. I think there wouldn't be anything relevant to your issue.

**Comment by 99bobster99** [comment]
Darn, I thought we were on to something there .. :(   

Then the code for `mpv.com` is using `CreateProcess` the same way I'm using it, except for the redirection I'm using.

I'll mess around with it a bit more, since I really like the way mpv handles flickering, it's like magic! I know vlc has an anti-flickering option, but I can't get it dialed in to be the same as mpv. :(

**Comment by avih** [comment]
Maybe you need to use `CreatePipe` - https://docs.microsoft.com/en-us/windows/win32/api/namedpipeapi/nf-namedpipeapi-createpipe

**Comment by 99bobster99** [comment]
Yeah, I'm using that already, the pipes are getting created successfully to and from ffmpeg. The weird thing is, is that I see the size of `mpv.exe` changing (in task manager) every time I write data to its stdin. I know its accepting the data, but it just isn't showing the video output window for me.

**Comment by avih** [comment]
Maybe you should look here for an example https://stackoverflow.com/questions/4310645/how-does-one-setup-a-pipe-between-two-child-processes-in-win32

**Comment by 99bobster99** [comment]
Good find Avih, I haven't seen that webpage yet! I would think that vlc would work if the read/write ends of the pipe were flipped. I'll check in the morning, my brain is mush, time for some sleep. Cheers Avih, thanks for assisting!! I'll let you know what I find out tomorrow.

**Comment by 99bobster99** [comment]
It seems others had the same issue from a certain point onward, what was changed and is it possible to change it back?

[(https://github.com/NHLGames/NHLGames/issues/253)]
